source/extensions/common/wasm/wasm_vm_base.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/common/wasm/wasm_vm.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+/**
+ * Wasm host stats.
+ */
+#define ALL_VM_STATS(COUNTER, GAUGE)                                                               \
+  COUNTER(created)                                                                                 \
+  COUNTER(cloned)                                                                                  \
+  GAUGE(active, NeverImport)
+
+struct VmStats {
+  ALL_VM_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+struct VmGlobalStats {
+  std::atomic<int> active_vms_{0};
+};
+
+// Wasm VM base instance. Provides common behavior (e.g. Stats).
+class WasmVmBase : public WasmVm {
+public:
+  WasmVmBase(Stats::ScopeSharedPtr scope, VmGlobalStats& global_stats, absl::string_view runtime)
+      : scope_(scope), runtime_prefix_(absl::StrCat("wasm_vm.", runtime, ".")),
+        global_stats_(global_stats),
+        stats_(VmStats{ALL_VM_STATS(POOL_COUNTER_PREFIX(*scope_, runtime_prefix_),
+                                    POOL_GAUGE_PREFIX(*scope_, runtime_prefix_))}),
+        runtime_(std::string(runtime)) {
+    stats_.created_.inc();
+    auto active = ++global_stats_.active_vms_;
+    stats_.active_.set(active);
+    ENVOY_LOG(debug, "WasmVm created {} now active", runtime_, active);
+  }
+  virtual ~WasmVmBase() {
+    auto active = --global_stats_.active_vms_;
+    // The stats referrd to by stats_ are resolved by name in the stats system.
+    stats_.active_.set(active);
```

Comment:
this looks race free now. Thanks!

Reply:
Sorry, I spoke too soon. I think this is not race free.

two WasmVmBase objects being destructed concurrently. The first one gets through line 47 leaving active at 1, and then loses the PC. The second one then runs line 47, so its local `active` is `0`. Then it calls `stats_.active_.set(0)`. Then the first thread regains control, and sets `stats_.active_.set(1)`.

I think this will be most obviously correct if you throw a mutex into VmGlobalStats and lock that while decrementing the global atomic and setting the stat.

Actually, can you remind me why we have the global atomic *and* the stat? Can we get by with one of them?

Reply:
Switching to a pure Gauge.  However, since Gauge::inc() does not return the new value there is still a race on the logging output, but that is not critical IMO. PTAL